src/ome_zarr_models/v04/multiscales.py

```diff
@@ -1,17 +1,144 @@
-from typing import Any
+from __future__ import annotations
+from collections import Counter
+from typing import Annotated, Any, Sequence, get_args
 
-from pydantic import Field, field_validator
+from pydantic import AfterValidator, Field, model_validator
 
 from ome_zarr_models.base import Base
-from ome_zarr_models.utils import _unique_items_validator
-from ome_zarr_models.v04.axes import Axis
+from ome_zarr_models.utils import duplicates
+from ome_zarr_models.v04.axes import Axis, AxisType
 from ome_zarr_models.v04.coordinate_transformations import (
     PathScale,
     PathTranslation,
     VectorScale,
     VectorTranslation,
 )
 from ome_zarr_models.v04.omero import Omero
+from pydantic_zarr.v2 import ArraySpec, GroupSpec
+import zarr
+
+VALID_NDIM = (2, 3, 4, 5)
+NUM_TX_MAX = 2
+
+
+def ensure_transform_dimensionality(
+    transforms: tuple[VectorScale | PathScale]
+    | tuple[VectorScale | PathScale, VectorTranslation | PathTranslation],
+) -> (
+    tuple[VectorScale | PathScale]
+    | tuple[VectorScale | PathScale, VectorTranslation | PathTranslation]
+):
+    """
+    Ensures that the elements in the input sequence define transformations with identical dimensionality.
+    """
+    if isinstance(transforms[0], PathScale) or (
+        len(transforms) > 1 and isinstance(transforms[1], PathTranslation)
+    ):
+        # it's not possible to check that the dimensionality of a path transform
+        # is the same as the dimensionality of a vector transform
+        return transforms
+
+    ndims = tuple(tx.ndim for tx in transforms)
+    ndims_set = set(ndims)
+    if len(ndims_set) > 1:
+        msg = (
+            "The transforms have inconsistent dimensionality. "
+            f"Got transforms with dimensionality = {ndims}."
+        )
+        raise ValueError(msg)
+    return transforms
```

Comment:
This doesn't actually change/mutate the transforms, so maybe get rid of the return?

Reply:
don't we need the `return` for it to work as a pydantic validator?